soufi/finders/yum.py

```diff
@@ -0,0 +1,192 @@
+                if self._test_url(repo.baseurl + loc):
+                    return repo.baseurl + loc
+        return None
+
+    def _walk_binary_repos(self, name):
+        packages = []
+        for repo_url in self.binary_repos:
+            repo = self._get_repo(repo_url)
+            if repo is None:
+                continue
+            for package in repo.findall(name):
+                # If we have a binary package matching our version, but with
+                # a different name, return the NVR fields
+                if self.version in (package.evr, package.vr):
+                    return self._nevra_or_none(package)
+                # Otherwise let's make it weird
+                packages.append(package)
+
+        # If we've made it here, things have gotten weird; this should be
+        # the case of a source RPM that produces binary RPMs with different
+        # names *and* versions (the lvm2 package from Red Hat is a good example
+        # of this).  In this case, we won't be able to make heads or tails of
+        # the response, unless (and only unless) it contains a single package.
+        try:
+            [package] = packages
+        except ValueError:
+            return None, None
+        return self._nevra_or_none(package)
+
+    def _nevra_or_none(self, package):
+        if package.sourcerpm == '':
+            # It's here, but has no sources defined!  Bummer...
+            return None, None
+        nevra = self._get_nevra(package.sourcerpm)
+        return nevra['name'], f"{nevra['ver']}-{nevra['rel']}"
+
+    # Cache repo downloads as they are slow and network-bound.
+    @classmethod
+    @functools.lru_cache(maxsize=128)
+    def _get_repo(cls, url):
+        try:
+            return repomd.load(url)
+        except urllib.error.HTTPError:
+            return None
+
+    @staticmethod
+    def _get_nevra(filename):
+        """Split out the NEVRA fields from an RPM filename."""
+
+        # It's easiest to do this by eating the filename backwards, dropping
+        # offset pointers as we go
+        if filename.endswith('.rpm'):
+            filename = filename[:-4]
+        arch_offset = filename.rfind('.')
+        arch = filename[1 + arch_offset :]
+        rel_offset = filename[:arch_offset].rfind('-')
+        rel = filename[1 + rel_offset : arch_offset]
+        ver_offset = filename[:rel_offset].rfind('-')
+        ver = filename[1 + ver_offset : rel_offset]
+        name = filename[:ver_offset]
+        # Sometimes the epoch is before the name, sometimes it's before the
+        # version.  Support both.
+        if ':' in ver:
+            epoch, ver = ver.split(':', maxsplit=1)
+        elif ':' in name:
+            epoch, name = name.split(':', maxsplit=1)
+        else:
+            epoch = ''
+        return dict(name=name, ver=ver, rel=rel, epoch=epoch, arch=arch)
+
+    # Do not cache these, as they swallow Timeout errors, and caching those
+    # results would be silly
```

Comment:
Wrap it in another method that doesn't catch the timeout!

Reply:
Though I am loathe to cache `stat(2)` calls for all of the obvious reasons, recent studies have shown that HTTP HEAD requests are, in fact, **NOT** `stat(2) calls!  😆 

An update to this to cache HTTP HEAD calls has been pushed.